File: src/main/java/com/appzonegroup/creditclub/pos/util/SocketJob.java
package com.appzonegroup.creditclub.pos.util;

import android.util.Log;

import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSocket;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;
import java.io.*;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.nio.ByteBuffer;
import java.security.KeyManagementException;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.security.cert.X509Certificate;

public class SocketJob {
    private static String TAG = SocketJob.class.getSimpleName();
    public static byte[] sslSocketConnectionJob(String host, int port, byte[] data) throws NoSuchAlgorithmException, KeyManagementException, IOException {
        byte[] messageByte = new byte[1000];
        boolean end = false;
        String dataString = "";

        TrustManager[] trustAllCerts = new TrustManager[]{
                new X509TrustManager() {
                    public X509Certificate[] getAcceptedIssuers() {
                        X509Certificate[] myTrustedAnchors = new X509Certificate[0];
                        return myTrustedAnchors;
                    }

                    public void checkClientTrusted(X509Certificate[] certs, String authType) {
                    }

                    public void checkServerTrusted(X509Certificate[] certs, String authType) {
                    }
                }
        };

        SSLContext sc = SSLContext.getInstance("TLS");
        sc.init(null, trustAllCerts, new SecureRandom());

        Log.d(TAG, "Connecting");
        SSLSocket sslsocket = (SSLSocket) sc.getSocketFactory().createSocket();
        sslsocket.connect(new InetSocketAddress(host, port), 30000);
        sslsocket.startHandshake();

        DataInputStream in = new DataInputStream(sslsocket.getInputStream());
        DataOutputStream out = new DataOutputStream(sslsocket.getOutputStream());



        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        //InputStream in = connectionSocket.getInputStream();

        //dialog.setMessage("Sending......");
        Log.d(TAG, "Sending......");
        Log.d(TAG, "Length to send " + data.length);
        byte [] outputInfo = TerminalUtils.appendLengthBytes(data);
        Log.d(TAG, "Bytes written to output " + new String(outputInfo));
        out.write(outputInfo);
        out.flush();
        //dialog.setMessage("Receiving....");
        Log.d(TAG, "Receiving....");
        int bytesRead;

        in.readFully(messageByte, 0, 2);
        System.out.println("Gotten length!!!!!");

        //ByteBuffer byteBuffer = ByteBuffer.wrap(messageByte, 0, 2);
        ByteBuffer byteBuffer = ByteBuffer.wrap(messageByte, 0, 2);

        int bytesToRead = byteBuffer.getShort();
        System.out.println("About to read " + bytesToRead + " octets");

        if(bytesToRead <= 1){
            return null;
        }

        //The following code shows in detail how to read from a TCP socket
        while(!end)
        {
            bytesRead = in.read(messageByte);
            baos.write(messageByte, 0, bytesRead);
            dataString += new String(messageByte, 0, bytesRead);
            if (dataString.length() == bytesToRead )
            {
                end = true;
            }
        }

        return baos.toByteArray();
    }


    public static byte[] socketConnectionJob(String host, int port, byte[] data) throws IOException {
        Socket connectionSocket;
        byte[] messageByte = new byte[1000];
        boolean end = false;
        String dataString = "";

        InetAddress serverAddr = InetAddress.getByName(host);
        Log.d(TAG, "Connecting");
        connectionSocket = new Socket(serverAddr, port);
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        //InputStream in = connectionSocket.getInputStream();
        DataInputStream in = new DataInputStream(connectionSocket.getInputStream());
        OutputStream out = connectionSocket.getOutputStream();
        //dialog.setMessage("Sending......");
        Log.d(TAG, "Sending......");
        Log.d(TAG, "Length to send " + data.length);
        out.write(TerminalUtils.appendLengthBytes(data));
        //dialog.setMessage("Receiving....");
        Log.d(TAG, "Receiving....");
        int bytesRead;


        messageByte[0] = in.readByte();
        messageByte[1] = in.readByte();
        //ByteBuffer byteBuffer = ByteBuffer.wrap(messageByte, 0, 2);
        ByteBuffer byteBuffer = ByteBuffer.wrap(messageByte, 0, 2);

        int bytesToRead = byteBuffer.getShort();
        System.out.println("About to read " + bytesToRead + " octets");
        //The following code shows in detail how to read from a TCP socket
        while(!end)
        {
            bytesRead = in.read(messageByte);
            baos.write(messageByte, 0, bytesRead);
            dataString += new String(messageByte, 0, bytesRead);
            if (dataString.length() == bytesToRead )
            {
                end = true;
            }
        }

        return baos.toByteArray();
    }
}

File: pos/src/main/java/com/appzonegroup/creditclub/pos/util/SocketJob.kt
package com.appzonegroup.creditclub.pos.util

import com.creditclub.pos.RemoteConnectionInfo
import okio.*
import java.io.Closeable
import java.io.IOException
import java.net.InetSocketAddress
import java.net.Socket
import java.nio.ByteBuffer
import java.security.KeyManagementException
import java.security.NoSuchAlgorithmException
import java.security.SecureRandom
import javax.net.ssl.SSLContext
import javax.net.ssl.SSLSocket
import javax.net.ssl.TrustManager

private var trustManagers: Array<TrustManager?>? = null

/**
 * @param remoteConnectionInfo
 * @param timeout in seconds
 */
class TcpClient(
    private val remoteConnectionInfo: RemoteConnectionInfo,
    private val timeout: Int = remoteConnectionInfo.timeout,
) : Closeable {
    private val isConnected: Boolean
        get() = socket?.isConnected == true

    private var socket: Socket? = null
    private var source: BufferedSource? = null
    private var sink: BufferedSink? = null

    @Throws(
        NoSuchAlgorithmException::class,
        KeyManagementException::class,
        IOException::class
    )
    fun sendAndReceive(
        data: ByteArray,
    ): ByteArray? {
        if (!isConnected) {
            connect()
        }

        val outputInfo = withPrependedLength(data)
        sink!!.write(outputInfo).flush()

        val sourceLength = source!!.readByteArray(2)
        val byteBuffer = ByteBuffer.wrap(sourceLength, 0, 2)
        val bytesToRead = byteBuffer.short.toLong()
        if (bytesToRead <= 1) {
            return null
        }

        val response = source!!.readByteArray(bytesToRead)
        recycle()
        return response
    }

    private fun recycle() {
        socket?.apply {
            close()
            source = null
            sink = null
        }
        socket = null
    }

    private fun connect() {
        if (isConnected) return

        socket = getSocket(
            remoteConnectionInfo = remoteConnectionInfo,
            timeout = timeout,
        )
        source = socket!!.source().buffer()
        sink = socket!!.sink().buffer()
    }

    override fun close() {
        socket?.apply {
            close()
            source = null
            sink = null
        }
    }
}

object SocketJob {
    @Throws(
        NoSuchAlgorithmException::class,
        KeyManagementException::class,
        IOException::class
    )
    fun execute(
        connectionInfo: RemoteConnectionInfo,
        data: ByteArray,
        timeout: Int = connectionInfo.timeout,
    ): ByteArray? = connectionInfo.tcp(timeout = timeout).use {
        it.sendAndReceive(data = data)
    }

    fun setTrustManagers(tManagers: Array<TrustManager?>) {
        trustManagers = tManagers
    }
}

fun RemoteConnectionInfo.tcp(
    timeout: Int = this.timeout,
) = TcpClient(
    remoteConnectionInfo = this,
    timeout = timeout,
)

private fun withPrependedLength(data: ByteArray): ByteArray {
    val dataLength = data.size.toShort()
    val destination = ByteArray(2 + dataLength)
    destination[0] = (dataLength.toInt() shr 8).toByte()
    destination[1] = dataLength.toByte()
    System.arraycopy(data, 0, destination, 2, dataLength.toInt())
    return destination
}

private fun getSocket(
    remoteConnectionInfo: RemoteConnectionInfo,
    timeout: Int,
): Socket {
    val timeoutInMilliSeconds = timeout * 1000
    if (!remoteConnectionInfo.sslEnabled) {
        return Socket(remoteConnectionInfo.host, remoteConnectionInfo.port).apply {
            soTimeout = timeoutInMilliSeconds
        }
    }

    val sc = SSLContext.getInstance("TLSv1.2")
    sc.init(null, trustManagers, SecureRandom())
    val sslSocket = sc.socketFactory.createSocket() as SSLSocket
    sslSocket.soTimeout = timeoutInMilliSeconds
    val address = InetSocketAddress(remoteConnectionInfo.host, remoteConnectionInfo.port)
    sslSocket.connect(address, timeoutInMilliSeconds)
    sslSocket.startHandshake()
    return sslSocket
}
